stability: panic on basictracer.errAssertionFailed

`gamma` cluster running sha 0b950ce (just upgraded)

Node 4 (`104.196.98.24`) died shortly after startup with:
```
panic: (*basictracer.errAssertionFailed) (0x1596ee0,0xc826278b40)

goroutine 90 [running]:
panic(0x1596ee0, 0xc826278b40)
        /usr/local/go/src/runtime/panic.go:481 +0x3e6
github.com/opentracing/basictracer-go.(*spanImpl).maybeAssertSanityLocked(0xc822fd38c0)
        /go/src/github.com/opentracing/basictracer-go/debug.go:27 +0xd5
github.com/opentracing/basictracer-go.(*spanImpl).Lock(0xc822fd38c0)
        /go/src/github.com/opentracing/basictracer-go/debug.go:19 +0x3b
github.com/opentracing/basictracer-go.(*spanImpl).Log(0xc822fd38c0, 0x0, 0x0, 0x0, 0x1a3f940, 0xe, 0x0, 0x0)
        /go/src/github.com/opentracing/basictracer-go/span.go:106 +0xaf
github.com/opentracing/basictracer-go.(*spanImpl).LogEvent(0xc822fd38c0, 0x1a3f940, 0xe)
        /go/src/github.com/opentracing/basictracer-go/span.go:94 +0x7d
github.com/cockroachdb/cockroach/util/log.Trace(0x7f953e264068, 0xc821a02ae0, 0x1a3f940, 0xe)
        /go/src/github.com/cockroachdb/cockroach/util/log/trace.go:29 +0x63
github.com/cockroachdb/cockroach/storage.(*Replica).processRaftCommand(0xc82062a360, 0xc826278b10, 0x8, 0x153db, 0x57a, 0x400000004, 0x1, 0x1457a96f6f42dd11, 0x0, 0x400000004, ...)
        /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1879 +0x5f2
github.com/cockroachdb/cockroach/storage.(*Replica).handleRaftReady(0xc82062a360, 0x0, 0x0)
        /go/src/github.com/cockroachdb/cockroach/storage/replica.go:1567 +0xbf7
github.com/cockroachdb/cockroach/storage.(*Store).processRaft.func1()
        /go/src/github.com/cockroachdb/cockroach/storage/store.go:2095 +0x35d
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker.func1(0xc8203596c0, 0xc820dabf90)
        /go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:139 +0x52
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunWorker
        /go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:140 +0x62
```

Latest log:
[cockroach.stderr.gz](https://github.com/cockroachdb/cockroach/files/311959/cockroach.stderr.gz)

--- a/storage/replica.go
+++ b/storage/replica.go
@@ -1894,6 +1894,10 @@ func (r *Replica) processRaftCommand(idKey storagebase.CmdIDKey, index uint64, r
 				// in the first place, leads to less legible code here). This code
 				// path is rare.
 				r.insertRaftCommandLocked(cmd)
+				// The client should get the actual execution, not this one. We
+				// keep the context (which is fine since the client will only
+				// finish it when the "real" incarnation applies).
+				cmd = nil
 			}
 		}
 		if forcedErr == nil {
@@ -1924,6 +1928,7 @@ func (r *Replica) processRaftCommand(idKey storagebase.CmdIDKey, index uint64, r
 
 	if cmd != nil {
 		cmd.done <- roachpb.ResponseWithError{Reply: br, Err: err}
+		close(cmd.done)
 	} else if err != nil && log.V(1) {
 		log.Errorc(r.context(context.TODO()), "error executing raft command: %s", err)
 	}

--- a/storage/replica_test.go
+++ b/storage/replica_test.go
@@ -5925,6 +5925,69 @@ func TestReplicaBurstPendingCommandsAndRepropose(t *testing.T) {
 	})
 }
 
+// TestReplicaDoubleRefurbish exercises a code path in which a command is seen
+// fit for refurbishment, but has already been refurbished earlier (with that
+// command being in-flight). See #7185.
+func TestReplicaDoubleRefurbish(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	var tc testContext
+	tc.Start(t)
+	defer tc.Stop()
+
+	var ba roachpb.BatchRequest
+	ba.Timestamp = tc.clock.Now()
+	ba.Add(&roachpb.PutRequest{Span: roachpb.Span{Key: roachpb.Key("r")}})
+	replica := *tc.rng.GetReplica()
+
+	// Make a Raft command; we'll set things up so that it will be considered
+	// for refurbishment multiple times.
+	tc.rng.mu.Lock()
+	cmd, err := tc.rng.prepareRaftCommandLocked(context.Background(), makeIDKey(), replica, ba)
+	ch := cmd.done // must not use cmd outside of mutex
+	tc.rng.mu.Unlock()
+
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	{
+		// Send some random request to advance the lease applied counter to
+		// make `cmd` refurbish when we put it into Raft.
+		var ba roachpb.BatchRequest
+		ba.Timestamp = tc.clock.Now()
+		pArgs := putArgs(roachpb.Key("foo"), []byte("bar"))
+		ba.Add(&pArgs)
+		if _, pErr := tc.rng.Send(context.Background(), ba); pErr != nil {
+			t.Fatal(pErr)
+		}
+	}
+
+	const num = 10
+	func() {
+		tc.rng.mu.Lock()
+		defer tc.rng.mu.Unlock()
+		// Insert the command and propose it ten times. Before the commit
+		// which introduced this test, the first application would repropose,
+		// and the second would decide to not repropose, but accidentally send
+		// the error to the client, so that the successful refurbishment would
+		// be the second result received by the client.
+		tc.rng.insertRaftCommandLocked(cmd)
+		for i := 0; i < num; i++ {
+			if err := tc.rng.proposePendingCmdLocked(cmd); err != nil {
+				t.Fatal(err)
+			}
+		}
+	}()
+
+	var i int
+	for resp := range ch {
+		i++
+		if i != 1 {
+			t.Fatalf("received more than one response on the done channel: %+v", resp)
+		}
+	}
+}
+
 // TestCommandTimeThreshold verifies that commands outside the replica GC
 // threshold fail.
 func TestCommandTimeThreshold(t *testing.T) {

